aip/client-libraries/4234.md

```diff
@@ -0,0 +1,103 @@
+approach to how individual `.proto` files are combined into targets, and to what those
+targets are named, clients can have a uniform experience from API to API.
+
+Some of the various language constraints are as follows:
+
+*   Java's `option java_outer_classname`: per the API style guide is the name of
+    the proto file, with "Proto" appended. Unless `option java_multiple_files` is true,
+    the file name defines the packaging of the types.
+*   Go's target names will define the import path for the result. However, Go's package
+    names have a more restrictive character set (e.g., no dashes), and in some cases, that
+    means that the import path and package name will disagree. This interferes with some
+    tooling.
+
+## Guidance
+
+### Targets
+
+In the following file name cases, if the service name or resource name is more than one
+word, they should be separated by underscores.
+
+#### Services and request/response messages
+
+Individual services along with their request and response messages **should** live in
+separate `proto_library` targets, named `{service}_service_proto`.
+
+**NOTE:** Services **should** not share request or response messages. Reuse of resources
+or common messages is preferred.
+    
+#### Resource messages
+
+*   Resources that are a primary resource for an individual service, **should** be 
+    in a `proto_library` target named `{resource}_proto`.
+*   Otherwise, per-service resources **should** be in a `proto_library` target named 
+    `{service}_resources_proto`.
+*   If there are remaining shared resources, they **should** be in a `proto_library`
+    target named `resources_proto`.
+    
+#### Additional messages
+
+Remaining necessary messages **should** be in a `proto_library` target named
+`common_proto`.
+
+### `.proto` file naming convention
+
+In the case where there's a single `.proto` file in the `srcs` of the
+`proto_library`, it should be named by replacing the `_proto` suffix with `.proto`.
+
+### `{lang}_proto_library` naming convention
+
+The per-language rules, e.g., `cc_proto_library`, **should** be named by replacing the
+`_proto` suffix on the `proto_library` target with `_{lang}_proto`. e.g.,
+
+```
+proto_library(name = "foo_proto", …)
+cc_proto_library(name = "foo_cc_proto", deps = [":foo_proto"], …)
+```
+
+### `{lang}_{rpc}_library` naming convention
+
+The per-language, per-RPC rules, e.g., `cc_grpc_library`, **should** be named by replacing
+the `_proto` suffix on the `{lang}_proto_library` target with `_{rpc}`. e.g.,
+
+```
+cc_proto_library(name = "bar_service_cc_proto", deps = [":bar_service_proto"], …)
+cc_grpc_library(
+    name = "bar_service_cc_grpc",
+    srcs = [":bar_service_proto"],
+    deps = [":bar_service_cc_proto"],
+    …
+)
+```
+
+### Adding services
+
+If in the course of an alpha or beta period for an API, one or more new services are added
+or resources are made shared or specific such that they'd belong to a different or new
```

Comment:
I am not sure what "shared or specific" means.

Reply:
They mean, respectively. transitioning between cases where a single resource was

- the only resource for the only RPC service, but now is used across multiple RPC services in the same API
- a shared resource for multiple RPC services, but was dropped out of the surface of several such that it becomes the only resource for a single RPC service.

In both cases, had you to distribute protos into files and `proto_library` targets, you'd get an alternative association. If you effected that change, you'd break clients who were importing them from the old locations, unless you take special pains to arrange for them to be available from multiple `proto_library` targets (old and new).

Reply:
Got it. However, since APIs are developed independently, is there really any way for a single API to know whether it is being used as a dependency anywhere?

Reply:
It is not. In the mono-repo, this was tractable with code indexing. In the world at large, we can't even put deprecation messages in the right places (e.g., if `old.proto` was still useful after the move, you can't label it with a deprecation notice). The hope, though, is that if this intra-version change is made, that consumers could tend to their own cleanup at their leisure, and then be ready to just change the version they consume rather than bundle a bunch of simultaneous refactoring with the version change.

To be more specific, say `v1alpha` emitted messages into `old` and then an administrative change said that this should have been in `new`. If `v1alpha` uses the `import public` logic to provide it in both, then

1. original `v1alpha` consumers can continue to use it
1. if they elect to, they can modify their code to use it the new "right" way
1. updated `v1alpha` users can use `v1beta` out of the box, even though `v1beta` dropped the messages out of `old`